### PR TITLE
Block store

### PR DESCRIPTION
**Pull request by cintel**
Add a BlockStore Component, update it to use non locking LMDB for diskstorage.
Update the BlockStore to track Transactions and Batches. 
Update BlockValidator and BlockPublisher to check transactions dependencies. 
Update BlockValidator to generate transaction delta lists on Fork switch and BlockPublisher to rebuild the pending transaction list when the chain head is updated.

### CODE REVIEW

**Comment by peterschwarz** [comment]
Were there any changes needed in how the `GenesisController` interacts with block store as a result of these changes?  Might explain the test failures.

File: validator/sawtooth_validator/journal/publisher.py [hunk 1]
```diff
@@ -169,12 +191,56 @@ def on_batch_received(self, batch):
         :return: None
         """
         with self._lock:
-            self._pending_batches.append(batch)
+            # first we check if the transaction dependencies are satisified
+            # The completer should have taken care of making sure all
+            # Batches containing dependent transactions were sent to the
+            # BlockPublisher proir to this Batch. So if there is a missing
```
Comment (agunde406) [inline validator/sawtooth_validator/journal/publisher.py]: typo "prior"

File: validator/sawtooth_validator/server/core.py [hunk 2]
```diff
@@ -75,13 +76,11 @@ def __init__(self, network_endpoint, component_endpoint, peer_list,
         merkle_db = LMDBNoLockDatabase(db_filename, 'n')
         context_manager = ContextManager(merkle_db)
 
-        block_db_filename = os.path.join(data_dir, 'block.lmdb')
+        block_db_filename = os.path.join(data_dir, 'block-{}.lmdb')
```
Comment (agunde406) [inline validator/sawtooth_validator/server/core.py]: Is this the file name suposed to be  "block-{}.lmdb" ?

File: validator/sawtooth_validator/journal/block_store.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    def __iter__(self):
+        return iter(self._block_store)
+
+    def __len__(self):
+        return len(self._block_store)
+
+    def __str__(self):
+        out = []
+        for v in self._block_store.values():
```
Comment (agunde406) [inline validator/sawtooth_validator/journal/block_store.py]: ".values()" does not seem to be defined for LMDBNoLockDatabase

File: validator/sawtooth_validator/journal/chain.py [hunk 4]
```diff
@@ -97,17 +97,46 @@ def _verify_block_signature(self, blkw):
         return signing.verify(blkw.block.header, blkw.block.header_signature,
                               blkw.header.signer_pubkey)
 
-    def _verify_block_batches(self, blkw):
+    def _verify_batches_dependencies(self, batch, committed_txn):
+        """Verify that all transactions dependencies in this batch have been
+        satisfied, ie already committed by this block or prior block in the
+        chain.
+
+        :param batch: the batch
```
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/chain.py]: Missing params in doc.  What is `committed_txn`?

File: validator/sawtooth_validator/journal/chain.py [hunk 5]
```diff
@@ -97,17 +97,46 @@ def _verify_block_signature(self, blkw):
...
+        :return:
+        Boolean: True if all dependencies are present.
+        """
+        for t in batch.transactions:
```
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/chain.py]: Suggest name more meaningful than single-character names for these.

File: validator/sawtooth_validator/journal/publisher.py [hunk 6]
```diff
@@ -140,24 +185,27 @@ def on_chain_updated(self, chain_head,
 
         chain_head: the new head of block_chain
         committed_batches:
-        uncommitted_batches: the list of transactions if any that are now
-            de-committed due to the switch in forks.
-
+        uncommitted_batches:
```
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/publisher.py]: Duplicate parameter doc?

File: validator/sawtooth_validator/journal/publisher.py [hunk 7]
```diff
@@ -90,8 +105,18 @@ def _build_block(self, chain_head):
...
+            self._validate_batch(batch)
+
+        # build the TransactionCache
+        self._committed_txn = TransactionCache(self._block_cache.block_store)
+        if chain_head.header_signature not in self._block_cache.block_store:
+            # if we opportunistically building a block
+            # we need to check make sure we track that blocks transactions
+            # as recorded.
+            for batch in chain_head.block.batches:
+                for t in batch.transactions:
```
Comment (peterschwarz) [inline validator/sawtooth_validator/journal/publisher.py]: t -> txn

**Comment by cintel** [comment]
@peterschwarz @agunde406 Please have a look all the updates are inplace.

File: validator/sawtooth_validator/journal/publisher.py [hunk 8]
```diff
@@ -65,14 +67,20 @@ def __init__(self,
...
         self._block_cache = block_cache
         self._state_view_factory = state_view_factory
         self._transaction_executor = transaction_executor
         self._block_sender = block_sender
-        self._pending_batches = []  # batches we are waiting for validation
-        self._validated_batches = []  # batches that are valid and can be added
-        # to the next block.
+        self._pending_batches = []  # batches we are waiting for validation,
+        # arranged in the order of batches received.
+        self._committed_txn = None  # Look-up cache for transactions that are
```
Comment (agunde406) [inline validator/sawtooth_validator/journal/publisher.py]: This is a duplicate. Defined above

Comment (cintel) [inline validator/sawtooth_validator/journal/publisher.py]: removed duplicated definition above.

**Pull request merged**